File: DotNet/Turmerik.Core/DriveExplorer/DriveItem.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Turmerik.Core.Helpers;

namespace Turmerik.Core.DriveExplorer
{
    public class DriveItem : DriveItem<DriveItem>
    {
        public DriveItem()
        {
        }

        public DriveItem(DriveItemCore src) : base(src)
        {
        }

        public DriveItem(DriveItem<DriveItem> src, int depth = 0) : base(src, depth)
        {
        }
    }

    public class DriveItem<TDriveItem, TData> : DriveItem<TDriveItem>
        where TDriveItem : DriveItem<TDriveItem, TData>
    {
        public DriveItem()
        {
        }

        public DriveItem(DriveItemCore src) : base(src)
        {
        }

        public DriveItem(DriveItem<TDriveItem, TData> src, int depth = 0) : base(src, depth)
        {
            Data = src.Data;
        }

        public DriveItem(DriveItem<TDriveItem> src, int depth = 0) : base(src, depth)
        {
        }

        public TData Data { get; set; }
    }
}

File: ParcelWs/libs/trmrk-filemanager-nglib/projects/trmrk-filemanager-nglib/src/lib/services/common/driveitems-manager-service/drive-item.ts
import { NullOrUndef } from '../../../../trmrk/core';
import { DriveEntryCore } from '../../../../trmrk/DotNetTypes/Turmerik.Core.FileManager.DriveEntry';

import { NamedItemCore } from '../indexedDb/core';

export enum DriveItemTypeCore {
  None = 0,
  Folder,
  File,
}

export interface ContentItemCore<TContent> extends NamedItemCore {
  Content: TContent;
}

export interface DriveItemCore<
  TDriveItem extends DriveItemCore<TDriveItem, TDriveItemType>,
  TDriveItemType = DriveItemTypeCore
> extends NamedItemCore,
    DriveEntryCore {
  itemType: TDriveItemType;
  childItems?: TDriveItem[] | NullOrUndef;
}

export interface DriveItem<TDriveItemType = DriveItemTypeCore, TData = any>
  extends DriveItemCore<DriveItem<TDriveItemType>, TDriveItemType> {
  data: TData;
}
